README

```diff
@@ -38,7 +39,7 @@ GCC.  However, the Predator plug-in itself is known to work with GCC 4.[5-9].x,
 (1) Install all dependences of Predator and Forester
 ----------------------------------------------------
    * C/C++ compiler
-     - Predator currently supports GCC 4.1.2+ and Clang 3.2+
```

Comment:
Is this being removed on purpose?

I am able to build Predator as a GCC plugin with `clang-3.4.2` and it seems to work properly.

Reply:
Yes, that part is split to two parts 'Building from sources against the system XXX'

Reply:
But this line does not say which compiler Predator can be loaded into.  It says which compiler Predator can be compiled *with*.  One can compile Predator using clang as the compiler and then load the resulting binary into GCC to run the analysis.

Reply:
Yes, I know. But for later version of clang it crash on dry-run regre test (self compilation) of Code Listener. I can't tell from what version it is going on. I have installed last three and it crashed on clang 6 (for gcc-plugin).

Reply:
I tried it now -- compiled Predator using `clang-6.0.1` against system `gcc-6.4.0` and it worked just fine:
```
$ CC=clang-6.0 CXX=clang++-6.0 ./switch-host-gcc.sh /usr/bin/gcc-6.4.0 
[...]
Trying to build Code Listener...

make: Entering directory '/home/kdudka/git/predator/cl'
'version_cl.h.tmp' -> 'version_cl.h'
mkdir -p ../cl_build
cd ../cl_build && cmake -D GCC_HOST='/usr/bin/gcc-6.4.0' ../cl
-- The C compiler identification is Clang 6.0.1
-- The CXX compiler identification is Clang 6.0.1
-- Check for working C compiler: /usr/lib/llvm/6/bin/clang-6.0
-- Check for working C compiler: /usr/lib/llvm/6/bin/clang-6.0 -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/lib/llvm/6/bin/clang++-6.0
-- Check for working CXX compiler: /usr/lib/llvm/6/bin/clang++-6.0 -- works
[...]
-- GCC_HOST: /usr/bin/gcc-6.4.0
-- GCC_INC_DIR: /usr/lib/gcc/x86_64-pc-linux-gnu/6.4.0/plugin/include
-- 'echo "#include <new>" | /usr/bin/gcc-6.4.0 -m32 -c -xc++ -' works
-- CC1_HOST: /usr/libexec/gcc/x86_64-pc-linux-gnu/6.4.0/cc1
-- checking whether fancy_abort() in /usr/libexec/gcc/x86_64-pc-linux-gnu/6.4.0/cc1 is mangled
-- GCC plug-ins will be built with C++ linkage
[...]
100% tests passed, 0 tests failed out of 191
[...]
100% tests passed, 0 tests failed out of 1152

Total Test time (real) =  14.48 sec
make: Leaving directory '/home/kdudka/git/predator/sl'
CC=clang-6.0 CXX=clang++-6.0 ./switch-host-gcc.sh /usr/bin/gcc-6.4.0: real 58.827, user 3:09.94, sys 16.400, cpu 350%
```

If you are still able to reproduce the crash, please report it as a separate issue with some steps to reproduce.

Reply:
Sorry, I believe you. It looks like another macos issue. The C/CXX compilers were AppleClang 10.0.0.10001044, Clang 6.0.1, Clang 7.0.1 and Clang 8.0.0. Compilation starts with many warnings from gcc plugin headers.

Is in gcc binary anything suspicious?

```
  3/191 Test   #3: compile-self-01-dry-run ..........***Failed    0.54 sec
[...]
/Users/nika/Documents/REPOS/repo-darwin/cl_build/tests/libcl_smoke_test.so: debug: plug-in is compiled against gcc 6.5.0, built at 20181026, conf: ../configure --build=x86_64-apple-darwin17.7.0 --prefix=/usr/local/Cellar/gcc@6/6.5.0 --libdir=/usr/local/Cellar/gcc@6/6.5.0/lib/gcc/6 --enable-languages=c,c++,objc,obj-c++,fortran --program-suffix=-6 --with-gmp=/usr/local/opt/gmp --with-mpfr=/usr/local/opt/mpfr --with-mpc=/usr/local/opt/libmpc --with-isl=/usr/local/opt/isl --with-system-zlib --enable-stage1-checking --enable-checking=release --enable-lto --with-build-config=bootstrap-debug --disable-werror --with-pkgversion='Homebrew GCC 6.5.0' --with-bugurl=https://github.com/Homebrew/homebrew-core/issues --disable-nls --disable-multilib
```


```
[...]
/Users/nika/Documents/REPOS/repo-darwin/cl/cl_dotgen.cc:226: debug: ClDotGenerator: created dot file '-dev-null' [internal location] [-fplugin=libcl_smoke_test.so]
/Users/nika/Documents/REPOS/repo-darwin/cl/cl_typedot.cc:200: debug: ClTypeDotGenerator: created dot file '/dev/null' [internal location] [-fplugin=libcl_smoke_test.so]
/Users/nika/Documents/REPOS/repo-darwin/cl_build/tests/libcl_smoke_test.so: debug: plug-in successfully initialized
/Users/nika/Documents/REPOS/repo-darwin/cl/cl_dotgen.cc:226: debug: ClDotGenerator: created dot file '-Users-nika-Documents-REPOS-repo-darwin-cl-gcc-clplug.c.dot' [internal location] [-fplugin=libcl_smoke_test.so]
/Users/nika/Documents/REPOS/repo-darwin/cl/cl_dotgen.cc:226: debug: ClDotGenerator: created dot file '-Users-nika-Documents-REPOS-repo-darwin-cl-gcc-clplug.c-safe_basename.dot' [internal location] [-fplugin=libcl_smoke_test.so]
/Users/nika/Documents/REPOS/repo-darwin/cl/cl_dotgen.cc:226: debug: ClDotGenerator: created dot file '-Users-nika-Documents-REPOS-repo-darwin-cl-gcc-clplug.c-init_plugin_name.dot' [internal location] [-fplugin=libcl_smoke_test.so]
/Users/nika/Documents/REPOS/repo-darwin/cl/gcc/clplug.c:1990: warning: 'EDGE_EH (exception edge)' not handled in 'handle_fnc_bb' [internal location]
/Users/nika/Documents/REPOS/repo-darwin/cl/gcc/clplug.c:1990: warning: 'EDGE_EH (exception edge)' not handled in 'handle_fnc_bb' [internal location]
/Users/nika/Documents/REPOS/repo-darwin/cl/gcc/clplug.c:1990: warning: 'EDGE_EH (exception edge)' not handled in 'handle_fnc_bb' [internal location]
/Users/nika/Documents/REPOS/repo-darwin/cl/gcc/clplug.c:1990: warning: 'EDGE_EH (exception edge)' not handled in 'handle_fnc_bb' [internal location]
/Users/nika/Documents/REPOS/repo-darwin/cl/gcc/clplug.c:1990: warning: 'EDGE_EH (exception edge)' not handled in 'handle_fnc_bb' [internal location]
/Users/nika/Documents/REPOS/repo-darwin/cl/gcc/clplug.c:1990: warning: 'EDGE_EH (exception edge)' not handled in 'handle_fnc_bb' [internal location]
/Users/nika/Documents/REPOS/repo-darwin/cl/gcc/clplug.c:1990: warning: 'EDGE_EH (exception edge)' not handled in 'handle_fnc_bb' [internal location]
In function 'void init_plugin_name(const plugin_name_args*)':
cc1plus: warning: unhandled GIMPLE_RESX
/Users/nika/Documents/REPOS/repo-darwin/cl/gcc/clplug.c:1839: note: raised from 'handle_stmt_resx' [internal location]
/Users/nika/Documents/REPOS/repo-darwin/cl/gcc/clplug.c:265:16: warning: unreachable label 'L11' [-fplugin=libcl_smoke_test.so]
/Users/nika/Documents/REPOS/repo-darwin/cl/gcc/clplug.c:279:1: warning: unreachable label 'L14' [-fplugin=libcl_smoke_test.so]
<unknown location>: warning: unreachable label 'L4' [-fplugin=libcl_smoke_test.so]
/Users/nika/Documents/REPOS/repo-darwin/cl/gcc/clplug.c:247:16: warning: unreachable label 'L5' [-fplugin=libcl_smoke_test.so]
/Users/nika/Documents/REPOS/repo-darwin/cl/gcc/clplug.c:247:16: warning: unreachable label 'L8' [-fplugin=libcl_smoke_test.so]
/Users/nika/Documents/REPOS/repo-darwin/cl/gcc/clplug.c:265:16: warning: unreachable label 'L11' [-fplugin=libcl_smoke_test.so]
/Users/nika/Documents/REPOS/repo-darwin/cl/gcc/clplug.c:279:1: warning: unreachable label 'L14' [-fplugin=libcl_smoke_test.so]
<unknown location>: warning: unreachable label 'L4' [-fplugin=libcl_smoke_test.so]
/Users/nika/Documents/REPOS/repo-darwin/cl/gcc/clplug.c:247:16: warning: unreachable label 'L5' [-fplugin=libcl_smoke_test.so]
/Users/nika/Documents/REPOS/repo-darwin/cl/gcc/clplug.c:247:16: warning: unreachable label 'L8' [-fplugin=libcl_smoke_test.so]
/Users/nika/Documents/REPOS/repo-darwin/cl/gcc/clplug.c:265:16: warning: unreachable label 'L11' [-fplugin=libcl_smoke_test.so]
/Users/nika/Documents/REPOS/repo-darwin/cl/gcc/clplug.c:279:1: warning: unreachable label 'L14' [-fplugin=libcl_smoke_test.so]
<unknown location>: warning: unreachable label 'L4' [-fplugin=libcl_smoke_test.so]
/Users/nika/Documents/REPOS/repo-darwin/cl/gcc/clplug.c:247:16: warning: unreachable label 'L5' [-fplugin=libcl_smoke_test.so]
/Users/nika/Documents/REPOS/repo-darwin/cl/gcc/clplug.c:247:16: warning: unreachable label 'L8' [-fplugin=libcl_smoke_test.so]
/Users/nika/Documents/REPOS/repo-darwin/cl/cl_dotgen.cc:226: debug: ClDotGenerator: created dot file '-Users-nika-Documents-REPOS-repo-darwin-cl-gcc-clplug.c-free_plugin_name.dot' [internal location] [-fplugin=libcl_smoke_test.so]
/Users/nika/Documents/REPOS/repo-darwin/cl/cl_dotgen.cc:226: debug: ClDotGenerator: created dot file '-Users-nika-Documents-REPOS-repo-darwin-cl-gcc-clplug.c-dummy_printer.dot' [internal location] [-fplugin=libcl_smoke_test.so]
In function 'void dummy_printer(const char*)':
cc1plus: warning: unhandled GIMPLE_NOP
/Users/nika/Documents/REPOS/repo-darwin/cl/gcc/clplug.c:1893: note: raised from 'cb_walk_gimple_stmt' [internal location]
/Users/nika/Documents/REPOS/repo-darwin/cl/cl_dotgen.cc:226: debug: ClDotGenerator: created dot file '-Users-nika-Documents-REPOS-repo-darwin-cl-gcc-clplug.c-trivial_printer.dot' [internal location] [-fplugin=libcl_smoke_test.so]
/Users/nika/Documents/REPOS/repo-darwin/cl/cl_dotgen.cc:226: debug: ClDotGenerator: created dot file '-Users-nika-Documents-REPOS-repo-darwin-cl-gcc-clplug.c-cl_warn.dot' [internal location] [-fplugin=libcl_smoke_test.so]
/Users/nika/Documents/REPOS/repo-darwin/cl/cl_dotgen.cc:226: debug: ClDotGenerator: created dot file '-Users-nika-Documents-REPOS-repo-darwin-cl-gcc-clplug.c-cl_error.dot' [internal location] [-fplugin=libcl_smoke_test.so]
/Users/nika/Documents/REPOS/repo-darwin/cl/cl_dotgen.cc:226: debug: ClDotGenerator: created dot file '-Users-nika-Documents-REPOS-repo-darwin-cl-gcc-clplug.c-type_db_hash.dot' [internal location] [-fplugin=libcl_smoke_test.so]
/Users/nika/Documents/REPOS/repo-darwin/cl/cl_dotgen.cc:226: debug: ClDotGenerator: created dot file '-Users-nika-Documents-REPOS-repo-darwin-cl-gcc-clplug.c-var_db_hash.dot' [internal location] [-fplugin=libcl_smoke_test.so]
*** WARNING *** there are active plugins, do not report this as a bug unless you can reproduce it without enabling any plugins.
Event                            | Plugins
PLUGIN_FINISH_UNIT               | libcl_smoke_test
PLUGIN_FINISH                    | libcl_smoke_test
PLUGIN_START_UNIT                | libcl_smoke_test
/Users/nika/Documents/REPOS/repo-darwin/cl/gcc/clplug.c: In function 'hashval_t var_db_hash(const void*)':
/Users/nika/Documents/REPOS/repo-darwin/cl/gcc/clplug.c:333:18: internal compiler error: Segmentation fault: 11
 static hashval_t var_db_hash(const void *p)
                  ^~~~~~~~~~~
```

Reply:
I got it. On macos there are 2 types of stdlib: libc   used by clang/llvm and libstdc   used by gcc. Which one is set depend on compiler. But if you have dependencies in your code like lib for gcc plugins or llvm passes, you must compile with the corresponding library.

gcc --> gcc [OK]
```
$ otool -L cl_build/tests/libcl_smoke_test.so
cl_build/tests/libcl_smoke_test.so:
	@rpath/libcl_smoke_test.so (compatibility version 0.0.0, current version 0.0.0)
	/usr/local/opt/gcc@6/lib/gcc/6/libstdc++.6.dylib (compatibility version 7.0.0, current version 7.22.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.0.0)
	/usr/local/lib/gcc/6/libgcc_s.1.dylib (compatibility version 1.0.0, current version 1.0.0)
```

gcc --> clang [BAD]
```
$ otool -L cl_build/tests/libcl_smoke_test.dylib
cl_build/tests/libcl_smoke_test.dylib:
	@rpath/libcl_smoke_test.dylib (compatibility version 0.0.0, current version 0.0.0)
	/usr/local/opt/gcc@6/lib/gcc/6/libstdc++.6.dylib (compatibility version 7.0.0, current version 7.22.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.0.0)
	/usr/local/lib/gcc/6/libgcc_s.1.dylib (compatibility version 1.0.0, current version 1.0.0)
```

clang --> clang [OK]
```
$ otool -L cl_build/tests/libcl_smoke_test.dylib
cl_build/tests/libcl_smoke_test.dylib:
	@rpath/libcl_smoke_test.dylib (compatibility version 0.0.0, current version 0.0.0)
	/usr/local/opt/llvm@7/lib/libc++.1.dylib (compatibility version 1.0.0, current version 1.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.0.0)
```

clang --> gcc [BAD]
```
$ otool -L cl_build/tests/libcl_smoke_test.so
cl_build/tests/libcl_smoke_test.so:
	@rpath/libcl_smoke_test.so (compatibility version 0.0.0, current version 0.0.0)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 400.9.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.0.0)
```
I'll upload the commit later.